Mathlib/Topology/CWComplex/Classical/Basic.lean

```diff
@@ -523,38 +523,220 @@ lemma CWComplex.cellFrontier_subset_finite_openCell [CWComplex C] (n : ℕ) (i :
+
+/-- Copy of a `Subcomplex` with a new `carrier` equal to the old one. Useful to fix definitional
+equalities. See Note [range copy pattern]. -/
+protected def copy (E : Subcomplex C) (F : Set X) (hF : F = E) : Subcomplex C :=
+  { carrier := F
+    I := E.I
+    closed' := hF.symm ▸ E.closed'
+    union' := hF.symm ▸ E.union' }
+
+@[simp] lemma coe_copy (E : Subcomplex C) (F : Set X) (hF : F = E) :
+  (E.copy F hF : Set X) = F := rfl
+
+lemma copy_eq (E : Subcomplex C) (F : Set X) (hF : F = E) : E.copy F hF = E :=
+  SetLike.coe_injective hF
+
+lemma union (E : Subcomplex C) :
+    D ∪ ⋃ (n : ℕ) (j : E.I n), openCell (C := C) (D := D) n j.1 = E := by
+  rw [E.union']
+  rfl
+
+lemma closed (E : Subcomplex C) : IsClosed (E : Set X) := E.closed'
+
+end Subcomplex
+
+end RelCWComplex
+
+namespace CWComplex
+
+export RelCWComplex (Subcomplex)
+
+namespace Subcomplex
+
+export RelCWComplex.Subcomplex (I closed union mem_carrier coe_eq_carrier mem ext copy coe_copy
+  copy_eq)
+
+end CWComplex.Subcomplex
+
+lemma CWComplex.Subcomplex.union {C : Set X} [CWComplex C] {E : Subcomplex C} :
+    ⋃ (n : ℕ) (j : E.I n), openCell (C := C) n j = E := by
+  have := RelCWComplex.Subcomplex.union E (C := C) (D := ∅)
+  rw [empty_union] at this
+  exact this
+
+/-- An alternative version of `Subcomplex.mk`: Instead of requiring that `E` is closed it requires
+  that for every cell of the subcomplex the corresponding closed cell is a subset of `E`. -/
+@[simps -isSimp]
+def RelCWComplex.Subcomplex.mk' [T2Space X] (C : Set X) {D : Set X} [RelCWComplex C D]
+    (E : Set X) (I : Π n, Set (cell C n))
+    (closedCell_subset : ∀ (n : ℕ) (i : I n), closedCell (C := C) n i ⊆ E)
+    (union : D ∪ ⋃ (n : ℕ) (j : I n), openCell (C := C) n j = E) : Subcomplex C where
+  carrier := E
+  I := I
+  closed' := by
+    have hEC : (E : Set X) ⊆ C := by
+      simp_rw [← union, ← union_iUnion_openCell_eq_complex (C := C)]
+      exact union_subset_union_right D
+        (iUnion_mono fun n ↦ iUnion_subset fun i ↦ by apply subset_iUnion_of_subset ↑i; rfl)
+    apply isClosed_of_isClosed_inter_openCell_or_isClosed_inter_closedCell hEC
+    · have : D ⊆ E := by
+        rw [← union]
+        exact subset_union_left
+      rw [inter_eq_right.2 this]
+      exact isClosedBase C
+    intro n _ j
+    by_cases h : j ∈ I n
+    · right
+      suffices closedCell n j ⊆ E by
+        rw [inter_eq_right.2 this]
+        exact isClosed_closedCell
+      exact closedCell_subset n ⟨j, h⟩
+    · left
+      rw [← union, union_inter_distrib_right]
+      apply IsClosed.union
+      · rw [inter_comm, (disjointBase n j).inter_eq]
+        exact isClosed_empty
+      · simp_rw [iUnion_inter]
+        suffices ⋃ m, ⋃ (i : ↑(I m)), openCell m (i : cell C m) ∩ openCell (C := C) n j = ∅ by
+          rw [this]
+          exact isClosed_empty
+        apply iUnion_eq_empty.2 fun m ↦ iUnion_eq_empty.2 fun i ↦ ?_
+        exact (disjoint_openCell_of_ne (by aesop)).inter_eq
```

Comment:
```suggestion
      suffices E ∩ openCell n j = ∅ by
        rw [this]
        exact isClosed_empty
      simp_rw [← union, union_inter_distrib_right, iUnion_inter]
      apply sup_eq_bot_iff.mpr ⟨(disjointBase n j).symm.inter_eq, ?_⟩
      apply iUnion_eq_empty.2 fun m ↦ iUnion_eq_empty.2 fun i ↦ ?_
      exact (disjoint_openCell_of_ne (by aesop)).inter_eq
```
(There is a bit of defeq abuse here, because we don't have `union_eq_empty` so one has to use `sup_eq_bot_iff` instead, but I'd say this is fine)